include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -696,6 +696,46 @@ __pattern_remove_if(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec,
     return {std::ranges::begin(__r) + __idx, std::ranges::begin(__r) + __n};
 }
 
+//------------------------------------------------------------------------
+// reverse
+//------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R>
+void
+__pattern_reverse(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _R&& __r)
+{
+    const auto __n = std::ranges::size(__r);
+    if (__n <= 1)
+        return;
+
+    oneapi::dpl::__par_backend_hetero::__parallel_for(_BackendTag{}, std::forward<_ExecutionPolicy>(__exec),
+                                                      unseq_backend::__reverse_functor<decltype(__n)>{__n}, __n / 2,
+                                                      std::forward<_R>(__r))
+        .__checked_deferrable_wait();
+}
+
+//------------------------------------------------------------------------
+// reverse_copy
+//------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _InRange, typename _OutRange>
+void
+__pattern_reverse_copy(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r)
+{
+    const auto __n = std::ranges::size(__in_r);
+    if (__n == 0)
```

Comment:
I think, this is incorrect. What if the size of `_OutRange` is also `0`? I mean it might be handled by `__parallel_for` call, but to me it's super unobvious.

I wonder if we should move all the checks to the API level, not to the pattern level. No strong opinion on that; I just wanted to express the idea

Reply:
> I think, this is incorrect. What if the size of _OutRange is also 0? I mean it might be handled by __parallel_for call, but to me it's super unobvious.

The API level (`__reverse_copy_fn`) creates views over the original ranges, trimmed to the common minimal size. If one range is empty, the other view is also empty.

Yes, it is not obvious. We can add asserts or comments - my preferable way. Moving that check to the API level comes with a drawback. For the host path, we simply transfer the processing to the iterator based patterns, which do that check. We do not want to repeat it.

Reply:
I've added an assert with a comment.